ENIGMAsystem/SHELL/Graphics_Systems/Direct3D9/DX9vertex.cpp

```diff
@@ -79,65 +177,54 @@ void vertex_color(int buffer, int color, double alpha) {
   enigma::vertexBuffers[buffer]->vertices.push_back(finalcol);
 }
 
-void vertex_submit(int buffer, int primitive, unsigned vertex_start, unsigned vertex_count) {
-  enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+void vertex_submit(int buffer, int primitive, unsigned start, unsigned count) {
+  const enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
   const enigma::VertexFormat* vertexFormat = enigma::vertexFormats[vertexBuffer->format];
 
   // this is fucking temporary until we rewrite the model classes and
   // figure out a proper way to flush
   d3dmgr->EndShapesBatching();
 
-  // if the contents of the vertex buffer are dirty then we need to update
-  // our native vertex buffer object "peer"
-  if (vertexBuffer->dirty) {
-    LPDIRECT3DVERTEXBUFFER9 vertexBufferPeer = NULL;
-    auto it = enigma::vertexBufferPeers.find(buffer);
-    size_t size = enigma_user::vertex_get_size(buffer);
+  enigma::graphics_prepare_vertex_buffer(buffer);
 
-    // if we have already created a native "peer" vbo for this user buffer,
-    // then we have to release it if it isn't big enough to hold the new contents
-    // or if it has just been frozen (so we can remove its D3DUSAGE_DYNAMIC)
-    if (it != enigma::vertexBufferPeers.end()) {
-      vertexBufferPeer = it->second;
+  size_t stride = 0;
+  LPDIRECT3DVERTEXDECLARATION9 vertexDeclaration = vertex_format_declaration(vertexFormat, stride);
+  d3dmgr->SetVertexDeclaration(vertexDeclaration);
 
-      D3DVERTEXBUFFER_DESC pDesc;
-      vertexBufferPeer->GetDesc(&pDesc);
+  LPDIRECT3DVERTEXBUFFER9 vertexBufferPeer = enigma::vertexBufferPeers[buffer];
+  d3dmgr->SetStreamSource(0, vertexBufferPeer, 0, stride);
 
-      if (size > pDesc.Size || vertexBuffer->frozen) {
-        vertexBufferPeer->Release();
-        vertexBufferPeer = NULL;
-      }
-    }
+  int primitive_count = enigma_user::draw_primitive_count(primitive, count);
 
-    // create either a static or dynamic vbo peer depending on if the user called
-    // vertex_freeze on the buffer
-    if (!vertexBufferPeer) {
-      d3dmgr->CreateVertexBuffer(
-        size, vertexBuffer->frozen ? D3DUSAGE_WRITEONLY : (D3DUSAGE_DYNAMIC | D3DUSAGE_WRITEONLY),
-        0, D3DPOOL_DEFAULT, &vertexBufferPeer, NULL
-      );
-      enigma::vertexBufferPeers[buffer] = vertexBufferPeer;
-    }
+  d3dmgr->DrawPrimitive(enigma::primitive_types[primitive], start, primitive_count);
 
-    // copy the vertex buffer contents over to the native peer vbo on the GPU
-    VOID* pVoid;
-    vertexBufferPeer->Lock(0, 0, (VOID**)&pVoid, D3DLOCK_DISCARD);
-    memcpy(pVoid, vertexBuffer->vertices.data(), size);
-    vertexBufferPeer->Unlock();
+  vertexDeclaration->Release();
+}
 
-    vertexBuffer->vertices.clear();
-    vertexBuffer->dirty = false;
-  }
+void index_submit(int buffer, int vertex, int primitive, unsigned start, unsigned count) {
```

Comment:
Let me verify the flow, here: this method is an implementation of a General method which is called from `General/GSVertex.cpp`, right? So all this logic is the DX11-specific implementation of moving the General vertex representation to the GPU?

Reply:
It's the basic one that the 3 overloads call, yes. And yes, that's exactly what it is, but it only moves it when necessary and it additionally draws the contents. `vertex_submit` also works the same way, hence how these index buffers are "modeled" after their design. But yeah, you basically understand how it works.